lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -771,24 +842,45 @@ void SingleExitLoopTransformer::ensureSingleExitBlock() {
 
     // Update loop info if this belongs to a parent loop.
     SILLoop *outsideBlockLoop = LI->getLoopFor(outsideBlock);
-    if (outsideBlockLoop != nullptr) {
-      // FIXME: We don't deal with cases where the nodes being moved in
-      // belong to another loop yet. e.g.,
+    if (outsideBlockLoop == nullptr) {
+      // outsideBlock is not part of any other loop. Simply add it to our loop.
+      loop->addBasicBlockToLoop(outsideBlock, LI->getBase());
+    } else {
+      // We deal with the case where the nodes being moved in
+      // belong to another loop. e.g.,
       // while ... {
       //   if ... {
       //     for(...) {...}
       //     break;
       //   }
       // }
-      // Check that `loop` is nested within `reachableLoop`.
-      assert(outsideBlockLoop->contains(loop) &&
-             "Nodes being moved belong to a non-nested loop.");
-      // Move the node into our loop.
-      outsideBlockLoop->removeBlockFromLoop(outsideBlock);
-      LI->changeLoopFor(outsideBlock, nullptr);
+      if (outsideBlockLoop->contains(loop)) {
+        // If our `loop` is nested within `outsideBlockLoop`.  Move the node
```

Comment:
is there test coverage on both this if branch, and the else below?

Reply:
Yes, we have coverage for both. The `if` branch was handled even before this PR (and is covered by the example with labeled breaks).

The else branch is the new behavior and the tests in this PR provide coverage.